File: lab4/7.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int val;
    Node* left;
    Node* right;
    Node(int v): val(v), left(nullptr), right(nullptr) {}
};

Node* insertNode(Node* node, int value) {
    if (!node) return new Node(value);
    if (value < node->val) node->left = insertNode(node->left, value);
    else if (value > node->val) node->right = insertNode(node->right, value);
    // дубликаты игнорируем, как в исходном коде
    return node;
}

int ans = 0;

int dfs(Node* u) {
    if (!u) return 0;
    int hl = dfs(u->left);
    int hr = dfs(u->right);
    ans = max(ans, hl + hr + 1);
    return max(hl, hr) + 1;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) return 0;
    Node* root = nullptr;
    for (int i = 0; i < n; ++i) {
        int v; cin >> v;
        root = insertNode(root, v);
    }
    dfs(root);
    cout << ans << "\n";
    return 0;
}

File: lab4/7.py
import sys
sys.setrecursionlimit(300000)


n = int(input())
ls = list(map(int, input().split()))

class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def insert(node, value):
    if node is None:
        return Node(value)
    if value < node.val:
        node.left = insert(node.left, value)
    elif value > node.val:
        node.right = insert(node.right, value)
    return node

root = None
for v in ls:
    root = insert(root, v)


ans = 0
def dfs(u):

    global ans
    if u is None:
        return 0
    hl = dfs(u.left)
    hr = dfs(u.right)
    if hl + hr + 1 > ans:
        ans = hl + hr + 1
    return max(hl, hr) + 1

dfs(root)
print(ans)